File: module-server/src/main/java/de/twomartens/wahlrecht/mapper/v1/ConstituencyMapper.java
package de.twomartens.wahlrecht.mapper.v1;

import de.twomartens.wahlrecht.model.db.Constituency;
import org.mapstruct.CollectionMappingStrategy;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.NullValueCheckStrategy;
import org.mapstruct.ReportingPolicy;

@Mapper(collectionMappingStrategy = CollectionMappingStrategy.ADDER_PREFERRED,
    nullValueCheckStrategy = NullValueCheckStrategy.ALWAYS,
    unmappedTargetPolicy = ReportingPolicy.IGNORE)
public interface ConstituencyMapper {

  de.twomartens.wahlrecht.model.dto.v1.Constituency mapToExternal(Constituency constituency);

  de.twomartens.wahlrecht.model.internal.Constituency mapToInternal(
      de.twomartens.wahlrecht.model.dto.v1.Constituency constituency);

  de.twomartens.wahlrecht.model.internal.Constituency mapToInternal(Constituency constituency);

  @Mapping(target = "id", ignore = true)
  @Mapping(target = "created", ignore = true)
  @Mapping(target = "lastModified", ignore = true)
  Constituency mapToDB(de.twomartens.wahlrecht.model.dto.v1.Constituency constituency);
}

File: module-server/src/main/kotlin/de/twomartens/wahlrecht/mapper/v1/ConstituencyMapper.kt
package de.twomartens.wahlrecht.mapper.v1

import de.twomartens.wahlrecht.model.db.Constituency
import org.mapstruct.*

@Mapper(
    collectionMappingStrategy = CollectionMappingStrategy.ADDER_PREFERRED,
    nullValueCheckStrategy = NullValueCheckStrategy.ALWAYS,
    unmappedTargetPolicy = ReportingPolicy.IGNORE
)
interface ConstituencyMapper {
    fun mapToExternal(constituency: Constituency): de.twomartens.wahlrecht.model.dto.v1.Constituency

    fun mapToInternal(constituency: de.twomartens.wahlrecht.model.dto.v1.Constituency): de.twomartens.wahlrecht.model.internal.Constituency

    fun mapToInternal(constituency: Constituency): de.twomartens.wahlrecht.model.internal.Constituency

    @Mapping(target = "id", ignore = true)
    @Mapping(target = "created", ignore = true)
    @Mapping(target = "lastModified", ignore = true)
    fun mapToDB(constituency: de.twomartens.wahlrecht.model.dto.v1.Constituency): Constituency
}